src/main/scala/org/deepdive/ddlog/DeepDiveLogParser.scala

```diff
@@ -73,41 +114,92 @@ class DeepDiveLogParser extends JavaTokenParsers {
     columnName ~ columnType ^^ {
       case(name ~ ty) => Column(name, ty)
     }
+
+  def CategoricalParser = "Categorical" ~> "(" ~> """\d+""".r <~ ")" ^^ { n => MultinomialType(n.toInt) }
+  def BooleanParser = "Boolean" ^^ { s => BooleanType }
+  def dataType = CategoricalParser | BooleanParser
+
   def schemaDeclaration: Parser[SchemaDeclaration] =
-    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ^^ {
-      case (r ~ isQuery ~ "(" ~ attrs ~ ")") => {
+    relationName ~ opt("?") ~ "(" ~ rep1sep(columnDeclaration, ",") ~ ")" ~ opt(dataType) ^^ {
+      case (r ~ isQuery ~ "(" ~ attrs ~ ")" ~ vType) => {
         val vars = attrs.zipWithIndex map { case(x, i) => Variable(x.name, r, i) }
         var types = attrs map { case(x) => x.t }
-        SchemaDeclaration(Attribute(r, vars, types), (isQuery != None))
+        val variableType = vType match {
+          case None => if (isQuery != None) Some(BooleanType) else None
+          case Some(s) => Some(s)
+        }
+        SchemaDeclaration(Attribute(r, vars, types), (isQuery != None), variableType)
       }
     }
 
+  def operator = "||" | "+" | "-" | "*" | "/" | "&" | "::"
+  val aggregationFunctions = Set("MAX", "SUM", "MIN", "ARRAY_ACCUM", "ARRAY_AGG")
+
+  // expressions
+  def expr : Parser[Expr] =
+    ( lexpr ~ operator ~ expr ^^ { case (lhs ~ op ~ rhs) => BinaryOpExpr(lhs, op, rhs) }
+    | lexpr
+    )
+
+  def lexpr : Parser[Expr] =
+    ( functionName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ { 
+        case (name ~ _ ~ args ~ _) => FuncExpr(name, args, (aggregationFunctions contains name))
+      }
+    | constant ^^ { ConstExpr(_) }
+    | variableName ^^ { VarExpr(_) }
+    | "(" ~> expr <~ ")" 
+    )
 
   // TODO support aggregate function syntax somehow
-  def cqHead = relationName ~ "(" ~ repsep(variableName, ",") ~ ")" ^^ {
+  def cqHead = relationName ~ "(" ~ rep1sep(expr, ",") ~ ")" ^^ {
       case (r ~ "(" ~ variableUses ~ ")") =>
         Atom(r, variableUses.zipWithIndex map {
-          case(name,i) => Variable(name, r, i)
+          case (e,i) => ColumnExpr(e, r, i)
```

Comment:
I wonder why we need to keep those redundant values with `ColumnExpr`.  If you carefully look at where `ColumnExpr` is used, you'll notice the `Atom` holding the `e`s is always available.  Hilariously, you have a lot of extra code that tries to maintain the weight of this useless extra layer all over the code, e.g., `variableToExpr`.  When extending the design and underlying types like we're doing now, please get excited to remove these weight and bloat from existing code rather than just blindly keeping things as they are.  Removing code is infinitely better than adding.  Code like playing TETRIS. :)